core/codenvy-hosted-machine-authentication-agent/src/main/java/com/codenvy/machine/authentication/agent/MachineLoginFilter.java

```diff
@@ -0,0 +1,134 @@
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.user.User;
+import org.eclipse.che.commons.user.UserImpl;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.ws.rs.core.HttpHeaders;
+import java.io.IOException;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+/**
+ * Protects user's machine from unauthorized access.
+ *
+ * @author Anton Korneta
+ */
+@Singleton
+public class MachineLoginFilter implements Filter {
+
+    private final String                 tokenServiceEndpoint;
+    private final HttpJsonRequestFactory requestFactory;
+
+    @Inject
+    public MachineLoginFilter(@Named("api.endpoint") String apiEndpoint,
+                              HttpJsonRequestFactory requestFactory) {
+        this.tokenServiceEndpoint = apiEndpoint + "/machine/token";
+        this.requestFactory = requestFactory;
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException,
+                                                                                                     ServletException {
+        final HttpServletRequest httpReq = (HttpServletRequest)request;
+        final HttpSession session = httpReq.getSession(false);
+        if (session != null) {
+            EnvironmentContext.getCurrent().setUser((User)session.getAttribute("principal"));
+            chain.doFilter(request, response);
+            return;
+        }
+        final String machineToken = extractToken(request);
+        if (isNullOrEmpty(machineToken)) {
+            ((HttpServletResponse)response).sendError(HttpServletResponse.SC_UNAUTHORIZED,
+                                                      "Authentication on machine failed, token is missed");
+            return;
+        }
+
+        try {
+            final UserDescriptor user = requestFactory.fromUrl(tokenServiceEndpoint + "/user/" + extractToken(request))
+                                                      .useGetMethod()
+                                                      .setAuthorizationHeader(machineToken)
+                                                      .request()
+                                                      .asDto(UserDescriptor.class);
+            final User machineUser = new UserImpl(user.getName(), user.getId(), machineToken, null, false);
+            EnvironmentContext.getCurrent().setUser(machineUser);
```

Comment:
I think you should unset user in finally block. Otherwise thread can be reused with preset user.